src/action-providers.ts

```diff
@@ -21,18 +21,87 @@ class RefactoringActionProvider implements vscode.CodeActionProvider {
     const ast = t.parse(document.getText());
     const selection = createSelectionFromVSCode(range);
 
-    return this.refactorings
-      .filter(refactoring => this.canPerform(refactoring, ast, selection))
-      .map(refactoring => this.buildCodeActionFor(refactoring));
+    const applicableRefactorings: RefactoringWithActionProvider[] = [];
+
+    const onCanPeform = (
+      path: t.NodePath<any>,
+      refactoring: RefactoringWithActionProvider
+    ) => {
+      if (isLegacyActionProvider(refactoring.actionProvider)) {
+        return;
+      }
+
+      if (refactoring.actionProvider.updateMessage) {
+        refactoring.actionProvider.updateMessage(path);
+      }
+
+      applicableRefactorings.push(refactoring);
+    };
+
+    t.traverseAST(ast, {
+      enter: (path: t.NodePath<any>) => {
+        this.refactorings.forEach(refactoring =>
+          this.canPerform(refactoring, path, selection, onCanPeform)
+        );
+      }
+    });
+
+    const applicableLegacyRefactorings = this.refactorings.filter(refactoring =>
+      this.canPerformLegacy(refactoring, ast, selection)
+    );
+
+    return [...applicableRefactorings, ...applicableLegacyRefactorings].map(
+      refactoring => this.buildCodeActionFor(refactoring)
+    );
   }
 
   private canPerform(
+    refactoring: RefactoringWithActionProvider,
+    path: t.NodePath<any>,
+    selection: Selection,
+    onCanPerform: (
+      matchedPath: t.NodePath<any>,
+      refactoring: RefactoringWithActionProvider
+    ) => void
+  ) {
```

Comment:
I find it surprising that `canPerform()` is not returning a boolean. I'm not sure I understand how the new implementation is supposed to determine which refactorings can actually be performed.

Isn't this PR breaking the Simplify Ternary quick fix?

Reply:
Yes, it is a bit strange that `canPerform()` calls a callback. However, moving the callback logic inside leads to `canPerform()` mutating the `refactoring` that is passed which is - to me - more surprising than the function's name. I will think about a different name which makes the function's intent more clear.